### PR TITLE
Side by side host config

### PR DESCRIPTION
**Pull request by acelaya**
This PR adds the logic needed to parse `sideBySide` config options and pass them to the `HTMLIntegration` by guest object.

It also documents the shape of the `sideBySide` config object.

### Testing steps

* With the feature flag enabled:
  1. Make sure you have `html_side_by_side` enabled for your user or all users in your local h instance: http://localhost:5000/admin/features
  2. Go to http://localhost:3000. The sidebar should be kept open if you click/interact with the main content.
  3. Now edit the `client-config.js.mustache` file, uncommenting the `sideBySide` config block
      ```diff
      - // sideBySide: {
      - //   mode: 'manual'
      - // },
      + sideBySide: {
      +   mode: 'manual'
      + },
      ```
  4. Repeat step 2, but this time, clicking/interacting with the main content should close the sidebar, which means side-by-side was "disabled".
* With the feature flag disabled:
  1. Make sure you have `html_side_by_side` disabled for your user or all users in your local h instance: http://localhost:5000/admin/features
  2. Go to http://localhost:3000. The sidebar should get closed if you click/interact with the main content.
  3. Now edit the `client-config.js.mustache` file, setting the `mode` to `auto`.
      ```diff
      sideBySide: {
      -   mode: 'manual'
      +   mode: 'auto'
      },
      ```
  4. Repeat step 2. The behavior should still be the same, as the feature flag takes precedence and does not allow to explicitly "enable" side-by-side.

> This PR is part of https://github.com/hypothesis/client/issues/5571

### CODE REVIEW

File: docs/publishers/config.rst [hunk 1]
```diff
@@ -450,6 +450,30 @@ loads.
...
+         }
+       };
+     };
+
+  The following keys are supported in the :option:`sideBySide` object.
+
+  .. option:: mode
+
+    ``auto`` or ``manual``. Auto is the default value, where the main page content will be adjusted based on the sidebar
+    status, while manual indicates the guest app wants to take full control of handling side-by-side, which will
```
Comment (robertknight) [inline docs/publishers/config.rst:474]:
```suggestion
    status, while manual indicates the web page wants to take full control of handling side-by-side, which will
```

The terms "guest" and "host" are used internally, but I don't think we use that terminology anywhere in our docs at present.

File: src/annotator/config/index.ts [hunk 2]
```diff
@@ -207,6 +207,10 @@ const configDefinitions: ConfigDefinitionMap = {
     defaultValue: null,
     getValue: getHostPageSetting,
   },
+  sideBySide: {
+    allowInBrowserExt: true,
```
Comment (acelaya) [inline src/annotator/config/index.ts:211]: I'm not sure if this option should be allowed in the browser extension or not. What's usually the logic to decide that?

File: docs/publishers/config.rst [hunk 3]
```diff
@@ -450,6 +450,30 @@ loads.
...
+    effectively disable that logic.
```
Comment (robertknight) [inline docs/publishers/config.rst:475]: It would be useful to provide some pointers here about how the web page is supposed to implement its own side-by-side logic (ie. using the `hypothesis:layoutchange`) event.

File: src/annotator/config/index.ts [hunk 4]
```diff
@@ -68,21 +77,12 @@ function configurationKeys(context: Context): string[] {
...
-    case 'profile':
-      return contexts.profile;
-    case 'all':
-      // Complete list of configuration keys used for testing.
-      return Object.values(contexts).flat();
-    default:
-      throw new Error(`Invalid application context used: "${context}"`);
+  if (context === 'all') {
+    // Complete list of configuration keys used for testing.
+    return Object.values(contexts).flat();
```
Comment (robertknight) [inline src/annotator/config/index.ts:82]: 👍

File: src/annotator/config/settings.ts [hunk 5]
```diff
@@ -99,16 +103,29 @@ export function settingsFrom(window_: Window): SettingsGetters {
...
+  function sideBySide(): SideBySideOptions {
+    const value = hostPageSetting('sideBySide');
+
+    return {
+      mode:
+        // The sideBySide prop has to be an object.
+        // The mode prop has to be 'auto' or 'manual', if defined.
+        !isObject(value) || !('mode' in value) || !isSideBySideMode(value.mode)
+          ? 'auto'
+          : value.mode,
```
Comment (robertknight) [inline src/annotator/config/settings.ts:115]: We could do with some more declarative way of expressing schemas and validating the config in future.

File: src/annotator/integrations/html-side-by-side.ts [hunk 6]
```diff
@@ -228,3 +229,7 @@ export function preserveScrollPosition(
 
   return scrollDelta;
 }
+
+export function isSideBySideMode(mode: unknown): mode is SideBySideMode {
+  return typeof mode === 'string' && ['auto', 'manual'].includes(mode);
```
Comment (robertknight) [inline src/annotator/integrations/html-side-by-side.ts:234]: The first clause of the `&&` is redundant, since the second will fail if `mode` isn't one of the strings in the array.

File: src/shared/is-object.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export function isObject(value: unknown): value is object {
+  return value !== null && typeof value === 'object' && !Array.isArray(value);
```
Comment (robertknight) [inline src/shared/is-object.ts:2]: A JavaScript array _is_ normally viewed as being an "object". Conceptually it is an object whose keys are all stringified numbers (`[1, 2, 3]` is `{ ["0"]: 1, ["1"]: 2, ["2"]: 3, length: 3 }`).

File: docs/publishers/config.rst [hunk 8]
```diff
@@ -450,6 +450,30 @@ loads.
...
 
+.. option:: sideBySide
+
+  ``Object``. This option lets you customize how the side-by-side mode behaves.
+
+  For example:
+
+  .. code-block:: javascript
+
+     window.hypothesisConfig = function () {
```
Comment (robertknight) [inline docs/publishers/config.rst:461]: We should probably start using modern JS (ie. arrow functions) in our code examples now.

File: docs/publishers/config.rst [hunk 9]
```diff
@@ -450,6 +450,30 @@ loads.
                               // relative from the receiving iframe.
     }
 
+.. option:: sideBySide
+
+  ``Object``. This option lets you customize how the side-by-side mode behaves.
```
Comment (robertknight) [inline docs/publishers/config.rst:455]: I think it would be useful to explain briefly what "side-by-side mode" means here.

**Review by robertknight — COMMENTED** [review]
I took a first pass over this. Generally looks good, though I suggested some updates to the docs.

File: src/annotator/config/index.ts (hunk 2 above)
Comment (robertknight) [reply src/annotator/config/index.ts:211]: We try to make a judgement call about which approach (ignoring or respecting the configuration) is likely to be most useful to a user of the browser extension. The intent of this configuration is about allowing websites to provide a better implementation of side-by-side than the default, so it probably does make sense to allow in the extension.

File: docs/publishers/config.rst (hunk 8 above)
Comment (acelaya) [reply docs/publishers/config.rst:461]: Right! I basically copy-pasted another example 😅

File: src/annotator/config/settings.ts (hunk 5 above)
Comment (acelaya) [reply src/annotator/config/settings.ts:115]: Yep. I think some validation library could be useful here.

File: src/annotator/integrations/html-side-by-side.ts (hunk 6 above)
Comment (acelaya) [reply src/annotator/integrations/html-side-by-side.ts:234]: Perhaps this is a case in which `includes(mode as string)` makes sense.

File: src/shared/is-object.ts (hunk 7 above)
Comment (acelaya) [reply src/shared/is-object.ts:2]: The question would be, when you are checking if something is an object, do you usually want arrays to be taken into consideration for your following logic?

They are effectively objects, but you usually check if something is an array in order to "iterate" over it or use array methods, while you check if something is an object in order to then check if it has certain type shape or property (which will not overlap with the array API in most of the cases).

But I see your point here. I don't have a hard opinion whether the array check should be there or not.

File: docs/publishers/config.rst (hunk 3 above)
Comment (acelaya) [reply docs/publishers/config.rst:475]: I was planning to link to that event's documentation, with a brief explanation, but I switch to something else and forgot.

File: src/shared/is-object.ts (hunk 7 above)
Comment (robertknight) [reply src/shared/is-object.ts:2]: > The question would be, when you are checking if something is an object, do you usually want arrays to be taken into consideration for your following logic?

I think this is going to depend on context. I probably would have chosen to push the `!Array.isArray()` check [down to the caller](https://github.com/inspect-js/is-object/issues/1#issuecomment-74918974). Out of interest I looked for the top N npm packages that implement an "is object" check as a function ([is-object](https://www.npmjs.com/package/is-object), [isobject](https://www.npmjs.com/package/isobject)) and they disagreed on this point.

File: src/annotator/integrations/html-side-by-side.ts (hunk 6 above)
Comment (robertknight) [reply src/annotator/integrations/html-side-by-side.ts:234]: > Perhaps this is a case in which includes(mode as string) makes sense

👍 - TS's error makes sense here, because passing a value of the "wrong" type to `includes` probably indicates an error most of the time, but we know what we're doing here.

File: src/shared/is-object.ts (hunk 7 above)
Comment (acelaya) [reply src/shared/is-object.ts:2]: Actually, the first one agrees with you https://github.com/inspect-js/is-object/blob/main/index.js.

I will remove that condition for now.

File: docs/publishers/config.rst [hunk 10]
```diff
@@ -465,7 +498,7 @@ These settings configure where the client's assets are loaded from.
    the URL where the contents of the hypothesis package are hosted, including
    the trailing slash. (Default: for production builds:
    ``"https://cdn.hypothes.is/hypothesis/X.Y.Z/"``, for development builds:
-   ``"http://localhost:3001/hypothesis/X.Y.Z/""`.
+   ``"http://localhost:3001/hypothesis/X.Y.Z/"``.
```
Comment (acelaya) [inline docs/publishers/config.rst:501]: Small unrelated typo fix.

File: docs/publishers/config.rst [hunk 11]
```diff
@@ -450,6 +450,39 @@ loads.
...
 
+.. option:: sideBySide
+
+  ``Object``. This option lets you customize how the side-by-side mode behaves.
+
+  .. note::
+
+    The side-by-side mode allows the space used by the web page's main content
+    area to be adapted while the sidebar is open, ensuring it does not overlap
+    with potentially annotatable content.
```
Comment (robertknight) [inline docs/publishers/config.rst:461]:
```suggestion
    with annotatable content.
```

"Potentially" seems superfluous here.

File: docs/publishers/config.rst [hunk 12]
```diff
@@ -450,6 +450,39 @@ loads.
                               // relative from the receiving iframe.
     }
 
+.. option:: sideBySide
+
+  ``Object``. This option lets you customize how the side-by-side mode behaves.
+
+  .. note::
+
+    The side-by-side mode allows the space used by the web page's main content
```
Comment (robertknight) [inline docs/publishers/config.rst:459]:
```suggestion
    Side-by-side mode allows the space used by the web page's main content
```

File: docs/publishers/config.rst [hunk 13]
```diff
@@ -450,6 +450,39 @@ loads.
...
+  .. option:: mode
+
+    ``auto`` or ``manual``. Auto is the default value, where the main page
+    content will be adjusted based on the sidebar status, while manual
+    indicates the web page wants to take full control of handling side-by-side,
+    which will effectively disable that logic.
+
+    When setting it to ``manual``, you probably want to
+    `listen for layout changes </publishers/events/#cmdoption-arg-hypothesis-layoutchange>`_
+    as well, so that you can adapt the main content area of the page.
```
Comment (robertknight) [inline docs/publishers/config.rst:484]: I think we should re-word this to indicate that web page authors _should_ respond to layout changes and adapt their content if that is reasonably possible, so as to provide an experience that is consistent with how Hypothesis works more generally. There will be cases where it is not reasonable, in which case the page author can just set the mode to "manual" and do nothing else.

Possible wording: "When setting this to `manual`, you should listen for layout changes and adapt content to fit alongside the sidebar, if it is reasonable to do so."

File: src/annotator/config/index.ts [hunk 14]
```diff
@@ -31,12 +31,21 @@ type ConfigDefinitionMap = Record<string, ConfigDefinition>;
  */
 type Context = 'sidebar' | 'notebook' | 'profile' | 'annotator' | 'all';
 
+function throwInvalidContext(context: string): never {
+  throw new Error(`Invalid application context used: "${context}"`);
+}
```
Comment (robertknight) [inline src/annotator/config/index.ts:36]: The `context` argument to `configurationKeys` always comes from us, right? It might be sufficient to just rely on TypeScript to validate that `configurationKeys` is always called with a legal `Context` value.

File: src/annotator/config/test/settings-test.js [hunk 15]
```diff
@@ -395,4 +395,47 @@ describe('annotator/config/settingsFrom', () => {
...
+      },
+      {
+        input: { mode: 'auto' },
+        expectedResult: { mode: 'auto' },
+      },
+      {
+        input: { mode: 'manual' },
+        expectedResult: { mode: 'manual' },
+      },
+    ].forEach(({ input, expectedResult }) => {
```
Comment (robertknight) [inline src/annotator/config/test/settings-test.js:421]: Coercing the input to a legal value makes sense at present, but it would be preferable to at least log a warning in future.

File: src/annotator/config/test/settings-test.js [hunk 16]
```diff
@@ -395,4 +395,47 @@ describe('annotator/config/settingsFrom', () => {
       });
     });
   });
+
+  describe('#sideBySide', () => {
+    [
+      {
+        input: 'foo',
+        expectedResult: { mode: 'auto' },
```
Comment (robertknight) [inline src/annotator/config/test/settings-test.js:403]: It is more common to use `expected` as the identifier for the expected return value, in cases where it is clear what this refers to.

File: src/annotator/integrations/test/html-side-by-side-test.js [hunk 17]
```diff
@@ -254,4 +255,19 @@ the fighting was.`;
       assert.equal(scrollRoot.scrollTop, initialScrollTop);
     });
   });
+
+  describe('isSideBySideMode', () => {
+    [
+      { mode: 'auto', expectedResult: true },
```
Comment (robertknight) [inline src/annotator/integrations/test/html-side-by-side-test.js:261]: I'd suggest to use just `expected` as the identifier here.

File: src/shared/test/is-object-test.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
+import { isObject } from '../is-object';
+
+describe('is-object', () => {
+  describe('isObject', () => {
+    [
+      { value: {}, expectedResult: true },
```
Comment (robertknight) [inline src/shared/test/is-object-test.js:6]: I'd suggest to use just `expected` as the identifier here.

File: src/shared/test/is-object-test.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
+import { isObject } from '../is-object';
+
+describe('is-object', () => {
```
Comment (robertknight) [inline src/shared/test/is-object-test.js:3]: The top-level describe block is not needed here, although some older tests might have them unnecessarily. We allow multiple top-level blocks in a test file, although very few happen to use that. For modules with a single export, there is normally one top-level block matching the export name.

File: docs/publishers/config.rst [hunk 20]
```diff
@@ -450,6 +450,39 @@ loads.
...
+         mode: 'manual'
+       }
+     });
+
+  The following keys are supported in the :option:`sideBySide` object.
+
+  .. option:: mode
+
+    ``auto`` or ``manual``. Auto is the default value, where the main page
+    content will be adjusted based on the sidebar status, while manual
```
Comment (robertknight) [inline docs/publishers/config.rst:478]:
```suggestion
    content will be automatically resized to fit alongside the sidebar, while manual
```

The meaning of "status" might be unclear here.

File: docs/publishers/config.rst [hunk 21]
```diff
@@ -450,6 +450,39 @@ loads.
...
+    indicates the web page wants to take full control of handling side-by-side,
+    which will effectively disable that logic.
```
Comment (robertknight) [inline docs/publishers/config.rst:480]:
```suggestion
    indicates the web page wants to take full control of handling side-by-side.
    This disables automatic resizing of the content.
```

I'm being more explicit about what "that logic" refers to.

File: docs/publishers/config.rst [hunk 22]
```diff
@@ -450,6 +450,39 @@ loads.
                               // relative from the receiving iframe.
     }
 
+.. option:: sideBySide
+
+  ``Object``. This option lets you customize how the side-by-side mode behaves.
```
Comment (robertknight) [inline docs/publishers/config.rst:455]:
```suggestion
  ``Object``. This option lets you customize how side-by-side mode behaves.
```

File: src/annotator/config/settings.ts [hunk 23]
```diff
@@ -99,16 +103,29 @@ export function settingsFrom(window_: Window): SettingsGetters {
     }
   }
 
+  function sideBySide(): SideBySideOptions {
+    const value = hostPageSetting('sideBySide');
+
+    return {
+      mode:
+        // The sideBySide prop has to be an object.
+        // The mode prop has to be 'auto' or 'manual', if defined.
```
Comment (robertknight) [inline src/annotator/config/settings.ts:112]: The code below seems clear enough to me that these comments can be omitted.

**Review by robertknight — APPROVED** [review]
LGTM with minor feedback on a few things that I'll leave you to handle.

File: src/annotator/config/index.ts (hunk 14 above)
Comment (acelaya) [reply src/annotator/config/index.ts:36]: Yep, I thought about that as well. There's room for improvement on type safety here. We could even try to define it in a way that the return type is more specific than `string[]` based on the value of the argument.

I will try to tackle this separately as soon as we have some room.

File: src/annotator/config/settings.ts (hunk 23 above)
Comment (acelaya) [reply src/annotator/config/settings.ts:112]: That's actually very true 😅 

There were more logic here at first, with comments and early returns, but I ended up extracting it to helper functions. Those have descriptive enough names so that the comments are indeed superfluous now.
